sdl2-vga-terminal/test/tests.cpp

```diff
@@ -172,71 +98,214 @@ TEST(VgaTerminal, moveCursorCircle)
 	term.moveCursorUp();
 	EXPECT_EQ(10, term.getX());
 	EXPECT_EQ(10, term.getY());
-	SDL_Quit();
 }
 
-TEST(VgaTerminal, moveCursorBorder)
+TEST(VgaTerminal, NoAutoScroll)
 {
-	ASSERT_EQ(0, SDL_Init(SDL_INIT_VIDEO));
-
-	std::string termTitle = "Hello Test";
+	std::string termTitle = "Test";
 	VgaTerminal term = VgaTerminal(termTitle, SDL_WINDOW_HIDDEN, -1, 0);
+	term.autoScroll = false;
+	term.writeXY(79, 24, "0", 7, 0);
+	EXPECT_EQ(79, term.getX());
+	EXPECT_EQ(24, term.getY());
 
-	term.gotoXY(79, 23);
+	term.writeXY(79, 20, "0", 7, 0);
+	EXPECT_EQ(0, term.getX());
+	EXPECT_EQ(21, term.getY());
+}
+
+class  SetViewportParameterTests : public ::testing::TestWithParam<std::tuple<int, int, int, int, bool, int, int>> {};
+TEST_P(SetViewportParameterTests, SetVieport)
+{
+	int vx = std::get<0>(GetParam());
+	int vy = std::get<1>(GetParam());
+	int vw = std::get<2>(GetParam());
+	int vh = std::get<3>(GetParam());
+	bool autoScroll = std::get<4>(GetParam());
+	int x = std::get<5>(GetParam());
+	int y = std::get<6>(GetParam());
+		
+	VgaTerminal term = VgaTerminal("SetViewport Test", SDL_WINDOW_HIDDEN, -1, 0);
+	term.autoScroll = autoScroll;
+
+	EXPECT_TRUE(term.setViewPort(vx, vy, vw, vh));
+	auto e = term.getViewport();
+	EXPECT_EQ(vx, e.x);
+	EXPECT_EQ(vy, e.y);
+	EXPECT_EQ(vw, e.w);
+	EXPECT_EQ(vh, e.h);
+	term.writeXY(x, y, "X", 15, 1);
+	auto tc = term.at(x, y);
+	EXPECT_EQ('X', tc.c);
+	EXPECT_EQ(15, tc.col);
+	EXPECT_EQ(1, tc.bgCol);
+}
+INSTANTIATE_TEST_SUITE_P(
+	VgaTerminal,
+	SetViewportParameterTests,
+	::testing::Values(
+		std::make_tuple(0, 0, 1, 1, false, 0, 0),    // SetViewport1
+		std::make_tuple(10, 10, 10, 10, true, 0, 0), // SetViewport2
+		std::make_tuple(5, 4, 20, 20, true, 5, 5)    // atViewport
+	)
+);
+
+class SetViewportNullErrTests : public ::testing::TestWithParam<std::tuple<int, int, int, int>> {};
+TEST_P(SetViewportNullErrTests, setViewportNullError)
+{
+	using ::testing::StartsWith;
+	using ::testing::EndsWith;
+
+	SDL_Rect vp;
+	vp.x = std::get<0>(GetParam());
+	vp.y = std::get<1>(GetParam());
+	vp.w = std::get<2>(GetParam());
+	vp.h = std::get<3>(GetParam());
+
+	VgaTerminal term = VgaTerminal("setViewportNullErr", SDL_WINDOW_HIDDEN, -1, 0);
+
+	testing::internal::CaptureStderr();
+	auto r = term.getViewport();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'r' is assigned a value that is never used.](https://app.codacy.com/manual/Raffaello/sdl2-vga-terminal/pullRequest?prid=5386869)